near-sdk/src/store/vec/mod.rs

```diff
@@ -0,0 +1,453 @@
+mod impls;
+mod iter;
+
+use borsh::{BorshDeserialize, BorshSerialize};
+use once_cell::unsync::OnceCell;
+
+use self::iter::{Iter, IterMut};
+use crate::collections::append_slice;
+use crate::utils::StableMap;
+use crate::{env, CacheEntry, EntryState, IntoStorageKey};
+
+const ERR_INCONSISTENT_STATE: &[u8] = b"The collection is an inconsistent state. Did previous smart contract execution terminate unexpectedly?";
+const ERR_ELEMENT_DESERIALIZATION: &[u8] = b"Cannot deserialize element";
+const ERR_ELEMENT_SERIALIZATION: &[u8] = b"Cannot serialize element";
+const ERR_INDEX_OUT_OF_BOUNDS: &[u8] = b"Index out of bounds";
+
+fn expect_consistent_state<T>(val: Option<T>) -> T {
+    val.unwrap_or_else(|| env::panic(ERR_INCONSISTENT_STATE))
+}
+
+/// An iterable implementation of vector that stores its content on the trie. This implementation
+/// will load and store values in the underlying storage lazily.
+///
+/// Uses the following map: index -> element. Because the data is sharded to avoid reading/writing
+/// large chunks of data, the values cannot be accessed as a contiguous piece of memory.
+///
+/// This implementation will cache all changes and loads and only updates values that are changed
+/// in storage after it's dropped through it's [`Drop`] implementation. These changes can be updated
+/// in storage before the variable is dropped by using [`Vector::flush`]. During the lifetime of
+/// this type, storage will only be read a maximum of one time per index and only written once per
+/// index unless specifically flushed.
+///
+/// This type should be a drop in replacement for [`Vec`] in most cases and will provide contracts
+/// a vector structure which scales much better as the contract data grows.
+///
+/// # Examples
+/// ```
+/// use near_sdk::store::Vector;
+///
+///# near_sdk::test_utils::test_env::setup();
+/// let mut vec = Vector::new(b"a");
+/// assert!(vec.is_empty());
+///
+/// vec.push(1);
+/// vec.push(2);
+///
+/// assert_eq!(vec.len(), 2);
+/// assert_eq!(vec[0], 1);
+///
+/// assert_eq!(vec.pop(), Some(2));
+/// assert_eq!(vec.len(), 1);
+///
+/// vec[0] = 7;
+/// assert_eq!(vec[0], 7);
+///
+/// vec.extend([1, 2, 3].iter().copied());
+/// assert!(Iterator::eq(vec.into_iter(), [7, 1, 2, 3].iter()));
+/// ```
+#[derive(BorshSerialize, BorshDeserialize)]
+pub struct Vector<T>
+where
+    T: BorshSerialize,
+{
+    len: u32,
+    prefix: Vec<u8>,
+    #[borsh_skip]
+    /// Cache for loads and intermediate changes to the underlying vector.
+    /// The cached entries are wrapped in a [`Box`] to avoid existing pointers from being
+    /// invalidated.
+    cache: StableMap<u32, OnceCell<CacheEntry<T>>>,
+}
+
+impl<T> Vector<T>
+where
+    T: BorshSerialize,
+{
+    /// Returns the number of elements in the vector, also referred to as its size.
+    pub fn len(&self) -> u32 {
```

Comment:
Consistency between targets imo, would be a footgun for some functionality to work in tests but then not work on chain. Happy to change to `usize`, but I'm just always hesitant to use `usize` in this type of context

Reply:
Yeah, consistency between targets is a good argument. It might make sense to put this into some kind of comment, maybe even a doc comment. As a user, I *would* wonder why we differ from standard rust APIs.